formats/json/commonMain/src/kotlinx/serialization/json/internal/TreeJsonEncoder.kt

```diff
@@ -69,7 +70,7 @@ private sealed class AbstractJsonTreeEncoder(
 
     override fun <T> encodeSerializableValue(serializer: SerializationStrategy<T>, value: T) {
         // Writing non-structured data (i.e. primitives) on top-level (e.g. without any tag) requires special output
-        if (currentTagOrNull != null || serializer.descriptor.kind !is PrimitiveKind && serializer.descriptor.kind !== SerialKind.ENUM) {
+        if (currentTagOrNull != null || !serializer.descriptor.needTopLevelTag) {
```

Comment:
It seems like it also fixes "improper" CONTEXTUAL usages reported https://github.com/Kotlin/kotlinx.serialization/issues/1774#issuecomment-977976962, right?

Reply:
Nope, it's better to reuse here `carrierDescriptor` then